fuse: Add cache for blob sizes

fd0 opened the pull request:
Closes: #820

fd0 committed:
fuse: Add cache for blob sizes

Closes: #820

middelink commented:
Should we run the tests with -race? maps aren't thread safe and I'm not 100% sure we cannot have multiple file listings in progress. aka, callbacks coming in from fuse.

Side note, this should be abstracted out, no? We now have multiple backends and cmd's where we are adding some form of caching to make things perform. Seems to me we should do this more central. For exmaple, I can imagine this commit not doing much on S3, as it already has a size cache...

fd0 commented:
Good point, but it's not relevant here:
 * Maps aren't thread safe, and multiple request from fuse can come in in parallel, that's both true. However, concurrent read from maps is thread safe, and we'll only create the size cache once (when the snapshots dir is constucted), and then it's read only access. So that's not an issue here.
 * The map caches sizes for all **blobs** in the repo, the data is extracted from the index. The size cache for the s3 backend caches the size of **files**. It's a different level of abstraction.

fd0 committed:
fuse: Improve semantics of the blob size cache

Wrap it in a struct and add a Lookup() function to make clear that it
is only queried, not changed, so we don't have any race conditions.

fd0 commented:
I've wrapped it in a struct and added a `Lookup()` method to make it clearer that the map is only queried, not changed concurrently.

fd0 committed:
Add entry to CHANGELOG

fd0 committed:
s3: Remove cache
